index.bs

```diff
@@ -90,24 +90,28 @@ A <dfn export id=file>file entry</dfn> additionally consists of
 a <dfn for="file entry">lock</dfn> (a string that may exclusively be "`open`", "`taken-exclusive`" or "`taken-shared`")
 and a <dfn for="file entry">shared lock count</dfn> (a number representing the number shared locks that are taken at a given point in time).
 
+The <dfn id="file-system-lock-queue">file system lock queue</dfn> is a
+[=parallel queue=] to be used for all [=tasks=] involving a [=file entry/lock=].
+
 <div algorithm>
 To <dfn for="file entry/lock">take</dfn> a [=file entry/lock=] with a |value| of "`exclusive`" or "`shared`" on a given [=file entry=] |file|:
 
 1. Let |lock| be the |file|'s [=file entry/lock=].
-1. Let |count| be the |file|'s [=file entry/shared lock count=].
-1. If |value| is "`exclusive`":
-   1. If |lock| is "`open`":
-     1. Set lock to "`taken-exclusive`".
-     1. Return true.
-1. If |value| is "`shared`":
-   1. If |lock| is "`open`":
-     1. Set |lock| to "`taken-shared`".
-     1. Set |count| to 1.
-     1. Return true.
-   1. Otherwise, if |lock| is "`taken-shared`":
-     1. Increase |count| by one.
-     1. Return true.
-1. Return false.
+1. [=Enqueue the following steps=] to the [=file system lock queue=]:
+  1. Let |count| be the |file|'s [=file entry/shared lock count=].
+  1. If |value| is "`exclusive`":
+     1. If |lock| is "`open`":
+       1. Set lock to "`taken-exclusive`".
+       1. Return true.
+  1. If |value| is "`shared`":
+     1. If |lock| is "`open`":
+       1. Set |lock| to "`taken-shared`".
+       1. Set |count| to 1.
+       1. Return true.
+     1. Otherwise, if |lock| is "`taken-shared`":
+       1. Increase |count| by one.
+       1. Return true.
+  1. Return false.
```

Comment:
This doesn't really work. These steps end up running in another conceptual thread. So the return values are lost.

You need to communicate back to the main thread through tasks.

You also need to actually start the parallel queue as otherwise it won't be executing its steps. See https://html.spec.whatwg.org/multipage/infrastructure.html#starting-a-new-parallel-queue.

(Note that the example there mistakenly doesn't queue tasks. That's a bug that still needs to be fixed.)

Reply:
Ah, I see now. I made the mistake of basing this off of the example :)

I've sketched out what lock acquisition + release could look like on a couple methods. Please let me know if there's a better way you'd recommend doing this. Once it looks good, I'll copy this pattern to all the relevant methods

Some thoughts:
- Instead of passing the promise directly, should the "take a lock" algorithm be passed an "onFailure" algorithm (which could then reject the promise)? Right now we always reject if the lock acquisition fails, but presumably we could add a method that doesn't
- Is there a better way to do this? I found https://github.com/whatwg/infra/issues/181, but that doesn't seem to have gone anywhere
- Is it correct to have `onLockReleasedAlgorithm` be an algorithm, or should it be a task?

Reply:
1. Yeah. Or a single algorithm that invoke with an argument to let it know success/failure.
2. Not currently.
3. No that's correct. The task is the eventual wrapper for the algorithm and some other state.

Reply:
Changed to pass only a single algorithm. Thanks for the suggestion!